components/BottomSheet/src/private/MDCSheetContainerView.m

```diff
@@ -0,0 +1,330 @@
+    return [self maximumSheetHeight];
+  }
+  return MIN(self.preferredSheetHeight, [self maximumSheetHeight]);
+}
+
+// Returns the maximum allowable height that the sheet can be dragged to.
+- (CGFloat)maximumSheetHeight {
+  CGSize size = self.bounds.size;
+  CGFloat scrollViewContentHeight = self.sheet.scrollView.contentSize.height;
+
+  // If we have a scrollview, the sheet should never get taller than its content height.
+  if (scrollViewContentHeight > 0) {
+    return MIN(size.height, scrollViewContentHeight);
+  } else {
+    return MIN(size.height, self.preferredSheetHeight);
+  }
+}
+
+#pragma mark - Gesture-driven animation
+
+- (void)animatePaneWithInitialVelocity:(CGPoint)initialVelocity {
+  self.sheetBehavior.targetPoint = [self targetPoint];
+  self.sheetBehavior.velocity = initialVelocity;
+  __weak MDCSheetContainerView *weakSelf = self;
+  self.sheetBehavior.action = ^{
+    [weakSelf sheetBehaviorDidUpdate];
+  };
+  [self.animator addBehavior:self.sheetBehavior];
+}
+
+// Calculates the snap-point for the view to spring to.
+- (CGPoint)targetPoint {
+  CGRect bounds = self.bounds;
+  CGFloat keyboardOffset = [MDCKeyboardWatcher sharedKeyboardWatcher].keyboardOffset;
+  CGFloat midX = CGRectGetMidX(bounds);
+  CGFloat bottomY = CGRectGetMaxY(bounds) - keyboardOffset;
+
+  switch(self.sheetState) {
+    case MDCSheetStatePreferred:
+      return CGPointMake(midX, bottomY - [self truncatedPreferredSheetHeight]);
+    case MDCSheetStateExtended:
+      return CGPointMake(midX, bottomY - [self maximumSheetHeight]);
+    case MDCSheetStateClosed:
+      return CGPointMake(midX, bottomY);
+  }
+}
+
+- (void)sheetBehaviorDidUpdate {
+  // If sheet has been dragged off the bottom, we can trigger a dismiss.
+  if (self.sheetState == MDCSheetStateClosed &&
+      CGRectGetMinY(self.sheet.frame) > CGRectGetMaxY(self.bounds)) {
+    [self.delegate sheetContainerViewDidHide:self];
+
+    [self.animator removeAllBehaviors];
+
+    // Reset the state to preferred once we are dismissed.
+    self.sheetState = MDCSheetStatePreferred;
+  }
+}
+
+#pragma mark - Notifications
+
+- (void)keyboardStateChangedWithNotification:(NSNotification *)notification {
+  if (self.window) {
+    [self animatePaneWithInitialVelocity:CGPointZero];
+  }
+}
+
+#pragma mark - MDCDraggableViewDelegate
+
+- (CGFloat)maximumHeightForDraggableView:(MDCDraggableView *)view {
+  return [self maximumSheetHeight];
+}
+
+- (BOOL)draggableView:(MDCDraggableView *)view shouldBeginDraggingWithVelocity:(CGPoint)velocity {
+  [self updateSheetState];
+
+  switch(self.sheetState) {
+    case MDCSheetStatePreferred:
+      return YES;
+    case MDCSheetStateExtended: {
+      UIScrollView *scrollView = self.sheet.scrollView;
+      if (scrollView) {
+        BOOL draggingDown = (velocity.y >= 0);
+        // Only allow dragging down if we are scrolled to the top.
+        if (scrollView.contentOffset.y <= -scrollView.contentInset.top && draggingDown) {
+          return YES;
```

Comment:
We already return YES for this switch case below.  Consider only coding for the exception:
if (scrollView.contentOffset.y > -scrollView.contentInset.top || !draggingDown) {
  return (CGRectGetHeight(scrollView.bounds) >= scrollView.contentSize.height);
}